e2e/deployment.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2021 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	deploymentutil "k8s.io/kubernetes/pkg/controller/deployment/util"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
+)
+
+// execCommandInPodWithName run command in pods using podName.
+func execCommandInPodWithName(
+	f *framework.Framework,
+	cmdString,
+	podName,
+	containerName,
+	nameSpace string) (string, string, error) {
+	cmd := []string{"/bin/sh", "-c", cmdString}
+	podOpt := framework.ExecOptions{
+		Command:            cmd,
+		PodName:            podName,
+		Namespace:          nameSpace,
+		ContainerName:      containerName,
+		Stdin:              nil,
+		CaptureStdout:      true,
+		CaptureStderr:      true,
+		PreserveWhitespace: true,
+	}
+
+	return f.ExecWithOptions(podOpt)
+}
+
+// loadAppDeployment loads the deployment app config and return deployment
+// object.
+func loadAppDeployment(path string) (*appsv1.Deployment, error) {
+	deploy := appsv1.Deployment{}
+	if err := unmarshal(path, &deploy); err != nil {
+		return nil, err
+	}
+
+	return &deploy, nil
+}
+
+// createDeploymentApp creates the deployment object and waits for it to be in
+// Available state.
+func createDeploymentApp(clientSet kubernetes.Interface, app *appsv1.Deployment, deployTimeout int) error {
+	_, err := clientSet.AppsV1().Deployments(app.Namespace).Create(context.TODO(), app, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to create deploy: %w", err)
+	}
+
+	return waitForDeploymentInAvailableState(app.Name, app.Namespace, clientSet, deployTimeout)
+}
+
+// deleteDeploymentApp deletes the deployment object.
+func deleteDeploymentApp(name, ns string, clientSet kubernetes.Interface, deployTimeout int) error {
```

Comment:
Consider using `clientSet` as 1st argument to all functions that need it. Some common ordering of arguments is pretty helpful when using these functions.

There are quite some functions already that make it user unfriendly. We should clean that up at one point, but that does not have priority. New functions, should probably use arguments in `framework/clientset, ..., timeout` order.

Reply:
Yes, make sense Niels. working on this now. Thanks!